### PR TITLE
Feat/add excluded connections to lattice connectivity

### PR DESCRIPTION
**Pull request by npoirie1**
Adds `excluded_connections` to `LatticeConnectivity`.

Issue #416

### CODE REVIEW

File: src/core/connectivity.jl [hunk 1]
```diff
@@ -213,20 +220,84 @@ struct LatticeConnectivity <: AbstractConnectivity
...
+                ),
+            )
+        end
+
+        num_qubits = nrows * ncols
+        if sorted_connection[2] > num_qubits
+            throw(
+                AssertionError(
+                    "connection $connection must have qubits with indices less than" *
+                    " $(num_qubits + 1)",
```
Comment (fSosaRey) [inline src/core/connectivity.jl:276]: theres no +1 here

File: src/core/connectivity.jl [hunk 2]
```diff
@@ -213,20 +220,84 @@ struct LatticeConnectivity <: AbstractConnectivity
...
+                ),
+            )
+        end
+
+        first_qubit_row_index = ((sorted_connection[1] - 1) ÷ ncols) + 1
```
Comment (fSosaRey) [inline src/core/connectivity.jl:281]: `mod( )` would be easier to parse than `÷` IMO

File: src/core/connectivity.jl (hunk 1 above)
Comment (npoirie1) [reply src/core/connectivity.jl:276]: If there are `n` qubits, the largest qubit index is `n`. The qubit indices therefore have to be less than `n + 1`.

File: src/core/connectivity.jl [hunk 3]
```diff
@@ -213,20 +220,84 @@ struct LatticeConnectivity <: AbstractConnectivity
...
+        near_index = sorted_connection[1] + ncols
+        coupler_exists =
+            sorted_connection[2] == near_index ||
+            (isodd(first_qubit_row_index) && sorted_connection[2] == near_index + 1) ||
```
Comment (fSosaRey) [inline src/core/connectivity.jl:285]: this logic only works for some LatticeDimensions and not others. Snowflurry.AnyonYamaskaConnectivity for instance does not follow this logic. Let's discuss how to address this offline

File: src/core/connectivity.jl (hunk 1 above)
Comment (fSosaRey) [reply src/core/connectivity.jl:276]: oh! right

File: src/core/connectivity.jl (hunk 3 above)
Comment (fSosaRey) [reply src/core/connectivity.jl:285]: changing the layout generation (https://github.com/SnowflurrySDK/Snowflurry.jl/pull/422) solves this problem

File: src/core/connectivity.jl [hunk 4]
```diff
@@ -264,7 +338,8 @@ with_excluded_positions(
 with_excluded_positions(
     c::LatticeConnectivity,
     excluded_positions::Vector{Int},
-)::LatticeConnectivity = LatticeConnectivity(c.qubits_per_row, excluded_positions)
+)::LatticeConnectivity =
+    LatticeConnectivity(c.qubits_per_row, excluded_positions, c.excluded_connections)
```
Comment (fSosaRey) [inline src/core/connectivity.jl:342]: is gonna break after rebasing, needs:
`LatticeConnectivity(c.dimensions[1], c.dimensions[2], excluded_positions, c.excluded_connections)`

File: src/core/connectivity.jl [hunk 5]
```diff
@@ -213,20 +220,84 @@ struct LatticeConnectivity <: AbstractConnectivity
...
     function LatticeConnectivity(
         qubits_per_row::Vector{Int},
         excluded_positions = Vector{Int}(),
+        excluded_connections = Vector{Tuple{Int,Int}}(),
     )
         nrows = maximum(qubits_per_row)
-        ncols = ceil(+(qubits_per_row...) / nrows)
+        ncols = ceil(Int, +(qubits_per_row...) / nrows)
+
+        sorted_connections =
```
[the remark this answers is not in the document]
Comment (npoirie1) [reply src/core/connectivity.jl:237]: Let's keep the sorted tuples.

File: src/core/connectivity.jl (hunk 3 above)
Comment (npoirie1) [reply src/core/connectivity.jl]: The logic now works with the new layout.

**Review by fSosaRey — APPROVED** [review]

**Pull request merged**